File: Baekjoon/Others/20208.cpp
#include<iostream>
#include<vector>
#define max(x, y) (x < y) ? y : x
using namespace std;

int answer, H;
vector<pair<int, int>> mint_choco;
vector<bool> visit;
pair<int, int> start;

int abs(int x) {
	return (x < 0) ? -x : x;
}

int dist(pair<int, int> a, pair<int, int> b) {
	return abs(a.first - b.first) + abs(a.second - b.second);
}

void dfs(int ind, int M, pair<int,int> last, int total) {
	if (ind) {
		if (dist(last, start) <= M) {
			answer = max(answer, total);
		}
	}
	if (ind == mint_choco.size()) return;

	for (int i = 0; i < mint_choco.size(); i++)
	{
		if (visit[i]) continue;
		if (M < dist(last, mint_choco[i])) continue;
		visit[i] = true;
		dfs(ind + 1, M + H - dist(last, mint_choco[i]), mint_choco[i], total + 1);
		visit[i] = false;
	}
}

int main() {
	//ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
	int N, M, i, j, tmp;
	cin >> N >> M >> H;
	for (i = 0; i < N; i++)
	{
		for (j = 0; j < N; j++)
		{
			cin >> tmp;
			if (tmp == 1) {
				start.first = i;
				start.second = j;
			}
			else if (tmp == 2) {
				mint_choco.push_back(make_pair(i, j));
			}
		}
	}
	for (i = 0; i < mint_choco.size(); i++)
	{
		visit.push_back(0);
	}
	dfs(0, M, start, 0);
	cout << answer << '\n';
	return 0;
}

File: Baekjoon/Others/20208.py
import sys; input = sys.stdin.readline
answer = 0
H = 0
visit = None
mint_choco = None
start = None

def distance(a, b):
    return abs(a[0]- b[0]) + abs(a[1] - b[1])

def dfs(ind, M, last, total):

    #현재 값이 최고인지 확인
    global answer
    if ind != 0:
        if distance(last, start) <= M:
            answer = max(answer, total)
    if ind == len(mint_choco):
        return

    for i in range(len(mint_choco)):
        if visit[i]: continue
        tmp_dist = distance(last, mint_choco[i])
        if M < tmp_dist: continue
        visit[i] = True
        dfs(ind + 1, M + H - tmp_dist, mint_choco[i], total + 1)
        visit[i] = False


def main():
    global visit, mint_choco, start, H
    N, M, H = map(int, input().split())
    mint_choco = []
    for i in range(N):
        tmp = list(map(int, input().split()))
        for j in range(N):
            if tmp[j] == 1:
                start = (i, j)
            elif tmp[j] == 2:
                mint_choco.append((i, j))
    visit = [False for _ in range(len(mint_choco))]
    # dfs(ind, M, last, total)
    dfs(0, M, start, 0)

    print(answer)


if __name__ == '__main__':
    main()